doc: dts: add guidance for modeling multi-API hardware (#105621)
zephyrproject-rtos/zephyr: doc/build/dts/design.rst

   / {
       chosen {
           zephyr,system-timer = &lptmr1;
       };
   };

The system-timer driver binds to the ``/chosen`` node. The counter
driver skips that instance and uses the remaining one.

**Per-instance software role with a property**

When no singleton system-wide selector applies, identical instances
can keep the same ``compatible`` and use a per-node property to
distinguish which API each instance provides.

Consider a SoC with three identical timer instances, where different boards or
applications may assign timer and counter roles independently:

.. code-block:: devicetree

   /* hardware description */
   timer0: timer@40000000 {
     compatible = "vendor,multitimer";
     reg = <0x40000000 0x1000>;
     /* ... */
   };

   timer1: timer@40001000 {
     compatible = "vendor,multitimer";
     reg = <0x40001000 0x1000>;
     /* ... */
   };

   timer2: timer@40002000 {
     compatible = "vendor,multitimer";
     reg = <0x40002000 0x1000>;
     /* ... */
   };

.. code-block:: devicetree

   /* integration-specific role assignment */
   &timer0 {
       zephyr,role = "counter";
   };

   &timer1 {
       zephyr,role = "pwm";
   };

   &timer2 {

>        zephyr,role = "counter";

Reviewer: `zephyr,role` seems to be identical to `device-type` or `device-api` or similar, a property describing the zephyr API to use, which has been discussed and rejected multiple times in this project.

A property like this, as described in this PR, is conflicting with the role of the `compatible`, given its the `compatible` which describes the hardware model for which a driver interacts with. That driver can expose any API it wants to the application/kernel, even no API.

The example above shows a device which can be both counter and pwm. They are a bit more similar than UART and I2C and SPI for example, which would not fit this model very well. For example:
```
multi_io0: multi-io@0 {
        compatible = "vendor,multi-io";
};

multi_io1: multi-io@1000 {
        compatible = "vendor,multi-io";
};

/* say we want the first one to be a SPI */
&multi_io0 {
        zephyr,role = "spi";
        #addr-cells = <1>; /* spi requires child node address */

        sensor@0 {
                compatible = "vendor,sensor";
        };
};

/* say we want the second one to be an UART */
&multi_io1 {
        zephyr,role = "uart";
        /* now no address cells are defined */
        baudrate = <115200>; /* this property is required for UART */

        sensor {
                compatible = "vendor,sensor";
        };
};
```
The properties required, like which bus something is, which phandle arrays are supported, which properties are required, change when the "role" of the node changes. This is because a different "role" means a different hardware model, thus its the `compatible` which needs to be changed.

In summary: The "role" is linked to how a driver will **use** some hardware, how a driver will **use** some hardware is a software property, not a hardware property. If the hardware model changes depending on the role, then the compatible needs to be changed, the role is not sufficient in these cases.

Reviewer 2: @bjarki-andreasen please read https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#the-zephyr-prefix and let us know if you still think this is an issue. I believe this is correctly using the zephyr, prefix and that is explicitly allowed in our use of DTS for software configuration for the reasons documented in that page.

Author: @bjarki-andreasen I agree with the distinction you are drawing here. If selecting the Zephyr-facing role also changes the DT-visible hardware model, then a per-node role property is the wrong tool; that case should use different compatible values or multifunction modeling instead.

I updated the guide to make that boundary explicit: the property-based guidance now only applies when all instances keep the same DT-visible binding schema, and I narrowed the example to a same-schema timer/counter case rather than a PWM/counter mix. The intent here is not to reintroduce device-api-style selection, but to document the narrower case where the hardware description is unchanged and only the per-instance software assignment varies. Singleton system-wide selection is still covered separately with /chosen.

******************************************************

When a hardware block can be used through different Zephyr APIs, devicetree
should keep hardware identity separate from the way the system uses each
instance.

Guidelines
==========

- Keep ``compatible`` focused on hardware identity. Avoid changing it only to
  route software API selection. A different ``compatible`` is appropriate when
  the hardware interfaces are architecturally distinct, for example because
  they use separate register blocks.

- Use ``/chosen`` for singleton system-wide selection. If the system must
  select exactly one instance for a global function, such as the system timer,
  console, or entropy source, represent that choice with a vendor-agnostic
  chosen property, even if other identical instances are used by other APIs.
  See also :ref:`devicetree-zephyr-chosen-nodes`.

- Use a property for per-instance software role assignment when the assignment
  must be expressed independently for each device instance and no singleton
  system-wide selector applies. In that case, keep the same ``compatible``
  and use a per-node property to describe that assignment. Use this only when

>   the DT-visible binding schema stays the same for all instances.

Reviewer: I disagree specifically with `Use a property for per-instance software role assignment`, primarily given the compatible can express the same thing, even if the hardware binding remains identical between two "software roles". Using the compatible is foundational to our device model, `_DT_INST_` macros are based on it for example.

I personally am not against adding a zephyr,role property, I myself have proposed variants of such a property, a property which selects which driver to use, and thus which API to expose. But that will need a discussion of its own, since it touches the device driver model as a whole, not just a few dt nodes.

Reviewer 3: > I disagree specifically with Use a property for per-instance software role assignment, primarily given the compatible can express the same thing

No it cannot express what Zephyr software API to use.

"_This key is used to match nodes to this binding._
...
_It should look like this in a binding file: compatible: "manufacturer,device"_
...
_The `device` part is usually from the datasheet._"
https://docs.zephyrproject.org/latest/build/dts/bindings-syntax.html#compatible

"_compatible: The name of the **hardware** device the node represents._
...
_The `device` part is usually taken from the datasheet._
...
_The compatible property can have multiple values. Additional values are useful when the device is a specific instance of a more general family_"
https://docs.zephyrproject.org/latest/build/dts/intro-syntax-structure.html#important-properties

Reviewer: > No it cannot express what Zephyr software API to use.

It can:

`vendor,device-counter` <- counter API
`vendor,device-pwm` <- pwm API

This is just as wrong as the zephyr,role property, but definitely doable, and has been used in zephyr already for rtc/counter bindings which today have been replaced by a virtual device `zephyr,rtc-counter`

Reviewer 3: > > No it cannot express what Zephyr software API to use.
> 
> It can:
> 
> `vendor,device-counter` <- counter API `vendor,device-pwm` <- pwm API

No, `counter` and `pwm` here (in compatible string) are supposed to refer to the hardware function of the device as described in datasheet, not Zephyr software APIs.

Reviewer: Sorry, I did not understand the comment.

When you say:

> No, counter and pwm here

what does "here" refer to? `zephyr,role = "pwm"` or `compatible = "vendor,device-pwm"`.

> are supposed to refer to the hardware function of the device as described in datasheet, not Zephyr software APIs.

This is the datasheet for Nordic SPI https://docs.nordicsemi.com/bundle/ps_nrf54L15/page/spim.html
This is the datasheet for Nordic TWIM (I2C) https://docs.nordicsemi.com/bundle/ps_nrf54L15/page/twim.html
This is the datasheet for Nordic UARTE https://docs.nordicsemi.com/bundle/ps_nrf54L15/page/uarte.html

These are all the same underlying hardware block which can be configured to support the various protocols. When the hardware is configured to be a SPI, it uses the compatible "nordic,nrf-spim". How would you model them with zephyr,role? what would it look like in devicetree, in drivers, in bindings, and finally in datasheets?

Reviewer 3: > When you say:
> 
> > No, counter and pwm here
> 
> what does "here" refer to? `zephyr,role = "pwm"` or `compatible = "vendor,device-pwm"`.

Of crs I am referring to compatible given what I am replying to.
 
> > are supposed to refer to the hardware function of the device as described in datasheet, not Zephyr software APIs.
> 
> This is the datasheet for Nordic SPI https://docs.nordicsemi.com/bundle/ps_nrf54L15/page/spim.html This is the datasheet for Nordic TWIM (I2C) https://docs.nordicsemi.com/bundle/ps_nrf54L15/page/twim.html This is the datasheet for Nordic UARTE https://docs.nordicsemi.com/bundle/ps_nrf54L15/page/uarte.html
> 
> These are all the same underlying hardware block which can be configured to support the various protocols. When the hardware is configured to be a SPI, it uses the compatible "nordic,nrf-spim". How would you model them with zephyr,role? what would it look like in devicetree, in drivers, in bindings, and finally in datasheets?

Answered in https://github.com/zephyrproject-rtos/zephyr/issues/105249#issuecomment-4067300413 and https://github.com/zephyrproject-rtos/zephyr/issues/105249#issuecomment-4242418868

For some reason, many people involved in this discussion appear to refuse to make the distinction between interfaces defined in hardware (HW interfaces, that are there regardless of what OS or HAL we are using) and interfaces introduced (invented) by implementing a given Zephyr public API (SW interfaces). The RFC and this PR are about the latter, and attempts to make the conversation about the former seem pointless to me.

When a hardware block can be used through different Zephyr APIs, devicetree
should keep hardware identity separate from the way the system uses each
instance.

Guidelines
==========

- Keep ``compatible`` focused on hardware identity. Avoid changing it only to
  route software API selection. A different ``compatible`` is appropriate when
  the hardware interfaces are architecturally distinct, for example because
  they use separate register blocks.

- Use ``/chosen`` for singleton system-wide selection. If the system must
  select exactly one instance for a global function, such as the system timer,
  console, or entropy source, represent that choice with a vendor-agnostic
  chosen property, even if other identical instances are used by other APIs.
  See also :ref:`devicetree-zephyr-chosen-nodes`.

- Use a property for per-instance role assignment when the assignment must be
  expressed independently on each instance and no singleton system-wide
  selector applies. In that case, keep the same ``compatible`` and use a
  per-node property to describe that assignment. Use this only when the
  instances still share the same DT-visible binding apart from that

>   assignment.

Reviewer 3: ```suggestion
  per-node property to describe that assignment.
```

  set of bindings supported by Linux.

- Devicetree source sharing between Zephyr and Linux is not done.

.. _dt-multi-api-hardware:

Modeling hardware exposed through multiple Zephyr APIs
******************************************************

When a hardware block can be used through different Zephyr APIs, devicetree
should keep hardware identity separate from the way the system uses each
instance.

Guidelines
==========

- Keep ``compatible`` focused on hardware identity. Avoid changing it only to
  route software API selection unless the hardware interfaces are
  architecturally distinct, for example because they use separate register
  blocks or bus protocols.

- Use ``/chosen`` for singleton system-wide selection. If the system must
  select exactly one instance for a global function, such as the system timer,
  console, or entropy source, represent that choice with a vendor-agnostic
  chosen property.

- Use a property for per-instance role assignment. If multiple identical
  instances serve different Zephyr APIs at the same time, keep the same

>   ``compatible`` and use a property such as ``role``, ``mode``, or

Reviewer 4: The existing [dts/bindings documentation](https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/build/dts/bindings-upstream.rst?plain=1#L255C49-L257C70) mentions that a prefix is needed. I'm not sure if using only `role` and `mode` here is enough - we need input from the DT maintainer @mbolivar  and bindings maintainer @JarmouniA  to clarify this. Additionally, if `zephyr,role` is a widely used property for distinguishing different Zephyr APIs, should we also document it?

  set of bindings supported by Linux.

- Devicetree source sharing between Zephyr and Linux is not done.

.. _dt-multi-api-hardware:

Modeling hardware exposed through multiple Zephyr APIs
******************************************************

When a hardware block can be used through different Zephyr APIs, devicetree
should keep hardware identity separate from the way the system uses each

> instance.

Reviewer 3: ```suggestion
should keep hardware identity separate from the way software drivers and subsystems use each
hardware instance.
```

Guidelines
==========

> - Keep ``compatible`` focused on the device programming model. Avoid changing

Reviewer 3: ```suggestion
- Keep ``compatible`` property focused on the device programming model. Avoid changing
```

  it only to route software API selection. A different ``compatible`` is
  appropriate when the hardware interfaces are architecturally distinct, for
  example because they use separate register blocks.

- Use ``/chosen`` for singleton system-wide selection. If the system must
  select exactly one instance for a global function, such as the system timer,

>   console, or entropy source, represent that choice with a vendor-agnostic

Reviewer 3: ```suggestion
  console, or entropy source, represent that choice with a platform-agnostic
```

>   chosen property, even if other identical instances are used by other APIs.

Reviewer 3: ```suggestion
  chosen property, even if other identical hardware instances are used by other APIs.
```

  See also :ref:`devicetree-zephyr-chosen-nodes`.

- Place the selection at the right integration level. If it is a silicon-level
  fixed attribute, describe it in ``soc.dtsi``. If it varies by board, product,
  or application, describe it in the board devicetree or an overlay.

If different uses of the hardware device have different required

> properties, child nodes, or bus semantics, then they do not share one binding

Reviewer 3: ```suggestion
properties, child nodes, or bus semantics, then they do not have to share one binding
```

.. _dt-multi-api-hardware:

Modeling hardware exposed through multiple Zephyr APIs
******************************************************

When a hardware block can be used through different Zephyr APIs, devicetree
should keep hardware identity separate from the way software drivers and
subsystems use each hardware instance.

Guidelines
==========

- Keep the ``compatible`` property focused on the device programming model.
  Avoid changing it only to route software API selection. A different
  ``compatible`` is appropriate when the hardware interfaces are
  architecturally distinct, for example because they use separate register
  blocks.

- Use ``/chosen`` for singleton system-wide selection. If the system must
  select exactly one instance for a global function, such as the system timer,
  console, or entropy source, represent that choice with a platform-agnostic
  chosen property, even if other identical hardware instances are used by
  other APIs. See also :ref:`devicetree-zephyr-chosen-nodes`.

- Place the selection at the right integration level. If it is a silicon-level
  fixed attribute, describe it in ``soc.dtsi``. If it varies by board, product,

>   or application, describe it in the board devicetree or an overlay.

Reviewer 5: Nits:
1. `soc.dtsi` is not a real file
2. use file extension for board too
```suggestion
- Place the selection at the right integration level. If it is a silicon-level
  fixed attribute, describe it in the SoC ``.dtsi``. If it varies by board, product,
  or application, describe it in the board ``.dts`` or an overlay.
```

Author: Updated.

  set of bindings supported by Linux.

- Devicetree source sharing between Zephyr and Linux is not done.

.. _dt-multi-api-hardware:

Modeling hardware exposed through multiple Zephyr APIs
******************************************************

When a hardware block can be used through different Zephyr APIs, devicetree
should keep hardware identity separate from the way the system uses each
instance.

Guidelines
==========

> - Keep ``compatible`` focused on hardware identity. Avoid changing it only to

Reviewer 6: I would use the Devicetree specification wording here, e.g. "programming model" instead of "identity".

Author: Updated.

  route software API selection. A different ``compatible`` is appropriate when
  the hardware interfaces are architecturally distinct, for example because
  they use separate register blocks.

- Use ``/chosen`` for singleton system-wide selection. If the system must
  select exactly one instance for a global function, such as the system timer,
  console, or entropy source, represent that choice with a vendor-agnostic
  chosen property, even if other identical instances are used by other APIs.
  See also :ref:`devicetree-zephyr-chosen-nodes`.

- Place the selection at the right integration level. If it is a silicon-level
  fixed attribute, describe it in ``soc.dtsi``. If it varies by board, product,
  or application, describe it in the board devicetree or an overlay.

> If different uses of the hardware device require different required

Reviewer 6: ```suggestion
If different uses of the hardware device have different required
```

Author: Updated.

properties, child nodes, or bus semantics, then they do not share one binding
schema. In that case, use different ``compatible`` values or model the device
as a multifunction one.

Per-instance software assignment for otherwise identical hardware, when no
singleton system-wide selector applies, remains a follow-up topic and needs a

> concrete in-tree driver example before it can be documented as guidance.

Reviewer 6: For people not familiar with the discussions going on here, this paragraph may seem rather confusing.

Author: Remove these lines.
